### PR TITLE
feat: enable retention period for warehouse uploads

### PR DESCRIPTION
**Pull request by cisse21**
# Description

Delete all the warehouse uploads which are beyond the retention period

## Linear Ticket

Fixes PIPE-1308

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

**Review by lvrach — APPROVED** [review]
non-blocking comments for discussion

File: warehouse/archive/cron.go [hunk 1]
```diff
@@ -15,7 +17,13 @@ func CronArchiver(ctx context.Context, a *Archiver) {
...
 				err := a.Do(ctx)
 				if err != nil {
-					a.log.Errorf(`Error archiving uploads: %v`, err)
+					a.log.Errorn(`Error archiving uploads`, logger.NewErrorField(err))
+				}
+			}
+			if a.config.canDeleteUploads.Load() {
+				err := a.Delete(ctx)
+				if err != nil {
+					a.log.Errorn(`Error deleting uploads`, logger.NewErrorField(err))
```
Comment (achettyiitr) [inline warehouse/archive/cron.go:26]:
```suggestion
					a.log.Errorn(`Error archiving uploads`, obskit.Error(err))
				}
			}
			if a.config.canDeleteUploads.Load() {
				err := a.Delete(ctx)
				if err != nil {
					a.log.Errorn(`Error deleting uploads`, obskit.Error(err))
```

File: warehouse/archive/archiver.go [hunk 2]
```diff
@@ -510,3 +512,79 @@ func (a *Archiver) deleteLoadFileRecords(
 
 	return nil
 }
+
+func (a *Archiver) Delete(ctx context.Context) error {
+	a.log.Infof(`[Archiver]: Started deleting for warehouse`)
```
Comment (achettyiitr) [inline warehouse/archive/archiver.go:517]: We can get rid of the `[Archiver]:` in logs.
```suggestion
	a.log.Infon(`Started deleting upload jobs`)
```

File: warehouse/archive/archiver_test.go [hunk 3]
```diff
@@ -214,6 +214,66 @@ func TestArchiver(t *testing.T) {
...
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	err = archiver.Delete(ctx)
+	require.NoError(t, err)
+
+	var count int
+	err = pgResource.DB.QueryRow(fmt.Sprintf(`SELECT COUNT(*) FROM %q`, warehouseutils.WarehouseUploadsTable)).Scan(&count)
+	require.NoError(t, err)
+	require.Equal(t, 0, count, "wh_uploads rows should be deleted")
```
Comment (achettyiitr) [inline warehouse/archive/archiver_test.go:274]:
```suggestion
	require.Zero(t, 0, count, "wh_uploads rows should be deleted")
```

**Review by achettyiitr — APPROVED** [review]

File: warehouse/archive/archiver.go [hunk 4]
```diff
@@ -510,3 +512,79 @@ func (a *Archiver) deleteLoadFileRecords(
...
+	}
+
+	maxLimit := a.config.maxLimit.Load()
+
+	for uploadsToDelete > 0 {
+		if err := a.deleteUploads(ctx, maxLimit); err != nil {
+			return fmt.Errorf("deleting uploads: %w", err)
+		}
+		uploadsToDelete -= maxLimit
+	}
```
[the remark this answers is not in the document]
Comment (cisse21) [reply warehouse/archive/archiver.go]: made the relevant changes

**Pull request merged**